compiler/src/iree/compiler/Dialect/Flow/Transforms/InsertDispatchDebugTargets.cpp

```diff
@@ -0,0 +1,253 @@
+      int localBreakOrdinal = -1;
+      if (fName == breakFname) localBreakOrdinal = breakOrdinal;
+      int localTraceOrdinal = -1;
+      if (fName == traceFname) localTraceOrdinal = traceOrdinal;
+
+      auto &bodyRegion = op.getFunctionBody();
+      auto dispatchOps = llvm::to_vector<8>(bodyRegion.getOps<DispatchOp>());
+
+      // Trace on a valid ordinal.
+      if (localTraceOrdinal > 0 && localTraceOrdinal < dispatchOps.size()) {
+        auto traceTarget = dispatchOps[localTraceOrdinal];
+        std::string entryPointName =
+            traceTarget.getEntryPoint().getRootReference().getValue().str();
+        for (FlatSymbolRefAttr nestedRef :
+             traceTarget.getEntryPoint().getNestedReferences()) {
+          entryPointName = (entryPointName + "::" + nestedRef.getValue()).str();
+        }
+        // Append the ordinal to the trace name.
+        traceOpWithName(traceTarget, entryPointName + std::string("::") +
+                                         std::to_string(localTraceOrdinal));
+      }
+
+      // Break on a valid ordinal, updating the function signature in the
+      // process. Currently only a single ordinal is supported so no need to
+      // check for overlapping breaks.
+      if (localBreakOrdinal > 0 && localBreakOrdinal < dispatchOps.size()) {
+        auto breakTarget = dispatchOps[localBreakOrdinal];
+        if (failed(replaceReturnWithOpResults(getOperation(), funcOp,
+                                              breakTarget)))
+          return signalPassFailure();
+      }
+    }
+  }
+};
+
+// Break/trace by symbol, after outlining dispatch regions and
+// deduplication.
+struct InsertDebugTargetAtSymbolPass
+    : public InsertDebugTargetAtSymbolBase<InsertDebugTargetAtSymbolPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Flow::FlowDialect, IREE::HAL::HALDialect>();
+  }
+  InsertDebugTargetAtSymbolPass(std::string breakStr, std::string traceStr) {
+    this->breakDebugTarget = breakStr;
+    this->traceDebugTarget = traceStr;
+  };
+  InsertDebugTargetAtSymbolPass(const InsertDebugTargetAtSymbolPass &pass)
+      : InsertDebugTargetAtSymbolPass(pass.breakDebugTarget,
+                                      pass.traceDebugTarget) {}
+
+  void runOnOperation() override {
+    for (auto it :
+         llvm::enumerate(getOperation().getOps<FunctionOpInterface>())) {
+      FunctionOpInterface funcOp = it.value();
+
+      DispatchOp breakTarget = DispatchOp();
+      funcOp.walk([&](DispatchOp dispatchOp) {
+        std::string entryPointName =
+            dispatchOp.getEntryPoint().getRootReference().getValue().str();
+        for (FlatSymbolRefAttr nestedRef :
+             dispatchOp.getEntryPoint().getNestedReferences()) {
+          entryPointName = (entryPointName + "::" + nestedRef.getValue()).str();
+        }
+        if (!traceDebugTarget.empty() &&
+            entryPointName.find(traceDebugTarget) != std::string::npos)
```

Comment:
we'll want regex to allow for things like saying `dispatch_1` but not `dispatch_10` +`dispatch_11` too - llvm::Regex should make this fairly easy (make Regex at top of runOnOperation and then `match` here and below)